File: 2021W35/HunRotation/pgr_84512.cpp
/*
    모음 사전
    https://programmers.co.kr/learn/courses/30/lessons/84512
*/

#include <string>
#include <vector>
#include <map>

using namespace std;

int solution(string word) {
    int answer = 0;
    vector<int> num_length = {1, 6, 31, 156, 781};
    map<char, int> vowels= {{'A', 0}, {'E', 1}, {'I', 2},
    {'O', 3}, {'U', 4}};

    int L = word.size();
    for(int i=0;i<L;i++) {
        answer += vowels[word[i]] * num_length[4-i] + 1;
    }

    return answer;
}

int main() {
    string str = "EIO";
    printf("%d", solution(str));
}

// keyword: 문자열
// 시간복잡도: O(word.size())
/*
idea: k개 이하 길이를 가진 단어의 개수를 미리 계산해 num_length로 사용
*/

File: 2021W35/fienestar/pgr_84512.js
/*
    5주차
    https://programmers.co.kr/learn/courses/30/lessons/84512
*/
// 6 ** 5 naive
function solution(word)
{
    let k = 0;
    let str = []
    function find()
    {   
        if(str.join('') === word)
            return k;
        
        if(str.length === 5){
            ++k;
            return;
        }
        
        ++k;
        
        for(const ch of "AEIOU") {
            str.push(ch)
            const result = find()
            if(result) return result;
            str.pop()
        }
    }
    
    return find()
}